### PR TITLE
[flang][acc] Add support for lowering combined constructs

### PR DESCRIPTION
**Pull request by razvanlupusoru**
PR#80319 added support to record combined construct semantics via an attribute. Add lowering support for this.

### CODE REVIEW

File: flang/lib/Lower/OpenACC.cpp [hunk 1]
```diff
@@ -2002,6 +2003,11 @@ createLoopOp(Fortran::lower::AbstractConverter &converter,
   if (!collapseDeviceTypes.empty())
     loopOp.setCollapseDeviceTypeAttr(builder.getArrayAttr(collapseDeviceTypes));
 
+  if (combinedConstructs) {
+    loopOp.setCombinedAttr(mlir::acc::CombinedConstructsTypeAttr::get(
+        builder.getContext(), combinedConstructs.value()));
```
Comment (clementval) [inline flang/lib/Lower/OpenACC.cpp:2008]:
```suggestion
        builder.getContext(), *combinedConstructs));
```

File: flang/lib/Lower/OpenACC.cpp [hunk 2]
```diff
@@ -2047,7 +2053,7 @@ genACC(Fortran::lower::AbstractConverter &converter,
       std::get<std::optional<Fortran::parser::DoConstruct>>(loopConstruct.t);
   auto loopOp = createLoopOp(converter, currentLocation, semanticsContext,
                              stmtCtx, *outerDoConstruct, eval, accClauseList,
-                             needEarlyExitHandling);
+                             {}, needEarlyExitHandling);
```
Comment (clementval) [inline flang/lib/Lower/OpenACC.cpp:2056]:
```suggestion
                             /*combinedConstructs=*/{}, needEarlyExitHandling);
```
Or you can put a default value with the argument

File: flang/lib/Lower/OpenACC.cpp [hunk 3]
```diff
@@ -2079,14 +2085,13 @@ static void genDataOperandOperationsWithModifier(
...
-                Fortran::lower::StatementContext &stmtCtx,
-                const Fortran::parser::AccClauseList &accClauseList,
-                bool outerCombined = false) {
+static Op createComputeOp(
+    Fortran::lower::AbstractConverter &converter,
+    mlir::Location currentLocation, Fortran::lower::pft::Evaluation &eval,
+    Fortran::semantics::SemanticsContext &semanticsContext,
+    Fortran::lower::StatementContext &stmtCtx,
+    const Fortran::parser::AccClauseList &accClauseList,
+    std::optional<mlir::acc::CombinedConstructsType> combinedConstructs) {
```
Comment (clementval) [inline flang/lib/Lower/OpenACC.cpp:2094]:
```suggestion
    std::optional<mlir::acc::CombinedConstructsType> combinedConstructs = std::nullopt) {
```

So we don't need to pass `{}` when there is not combined construct.

File: flang/lib/Lower/OpenACC.cpp [hunk 4]
```diff
@@ -2681,18 +2691,18 @@ genACC(Fortran::lower::AbstractConverter &converter,
   if (blockDirective.v == llvm::acc::ACCD_parallel) {
     createComputeOp<mlir::acc::ParallelOp>(converter, currentLocation, eval,
                                            semanticsContext, stmtCtx,
-                                           accClauseList);
+                                           accClauseList, {});
```
Comment (clementval) [inline flang/lib/Lower/OpenACC.cpp:2694]:
```suggestion
                                           accClauseList, /*combinedConstructs=*/{});
```
IF you want to keep it this way

File: flang/lib/Lower/OpenACC.cpp [hunk 5]
```diff
@@ -2681,18 +2691,18 @@ genACC(Fortran::lower::AbstractConverter &converter,
...
   } else if (blockDirective.v == llvm::acc::ACCD_data) {
     genACCDataOp(converter, currentLocation, eval, semanticsContext, stmtCtx,
                  accClauseList);
   } else if (blockDirective.v == llvm::acc::ACCD_serial) {
     createComputeOp<mlir::acc::SerialOp>(converter, currentLocation, eval,
                                          semanticsContext, stmtCtx,
-                                         accClauseList);
+                                         accClauseList, {});
```
Comment (clementval) [inline flang/lib/Lower/OpenACC.cpp:2701]:
```suggestion
                                         accClauseList, /*combinedConstructs=*/{});
```

File: flang/lib/Lower/OpenACC.cpp [hunk 6]
```diff
@@ -2681,18 +2691,18 @@ genACC(Fortran::lower::AbstractConverter &converter,
...
   } else if (blockDirective.v == llvm::acc::ACCD_kernels) {
     createComputeOp<mlir::acc::KernelsOp>(converter, currentLocation, eval,
                                           semanticsContext, stmtCtx,
-                                          accClauseList);
+                                          accClauseList, {});
```
Comment (clementval) [inline flang/lib/Lower/OpenACC.cpp:2705]:
```suggestion
                                          accClauseList, /*combinedConstructs=*/{});
```

**Review by clementval — COMMENTED** [review]
Couple of nits. 

If you want to do a stack PR you can open the first PR on a branch on the llvm-project repo and then make this PR against the other branch.

**Comment by razvanlupusoru** [comment]
Woah - I don't know how I caused this mess on rebase - please ignore this. Sorry to all of the implicit reviewers.

**Pull request closed without merging**